invadedhouse/commands/command_interpreter.py

```diff
@@ -0,0 +1,82 @@
+from .attack import Attack
+from .drop import Drop
+from .enter import Enter
+from .exit import Exit
+from .help import Help
+from .inventory import Inventory
+from .look import Look
+from .lookat import LookAt
+from .loot import Loot
+from .pickup import Pickup
+from .rename import Rename
+from .smash import Smash
+from .startover import StartOver
+from .warp import Warp
+from .whereto import WhereTo
+
+from models import House
+from models import Player
+
+
+class CommandInterpreter:
+    """
+    Processes commands input by player at command prompt.
+    """
+    def __init__(self, player: Player, house: House):
+        self.player = player
+        self.house = house
+
+        help_command = Help('help', player, house)
+        self.commands = [
+            Attack('attack', player, house),
+            Drop('drop', player, house),
+            Enter('enter', player, house),
+            Exit('exit', player, house),
+            help_command,
+            Inventory('inventory', player, house),
+            Look('look', player, house),
+            LookAt('look at', player, house),
+            Loot('loot', player, house),
+            Pickup('pickup', player, house),
+            Rename('change my name', player, house),
+            Smash('smash', player, house),
+            StartOver('start over', player, house),
+            Warp('warp', player, house),
+            WhereTo('where to', player, house)
+        ]
+        help_command.set_all_commands(self.commands)
+
+    def run(self):
+        while True:
+            cmd = self.get_next_command()
+            self.interpret_command(cmd)
+
+    @staticmethod
+    def get_next_command() -> str:
+        cmd = input('> ')
+        return cmd
+
+    def interpret_command(self, command_str: str):
+        command_str = command_str.lower()
+        words = command_str.split()
+        matched_command_str = ''
+        text = ''
+        matched_command = None
+
+        for word in words:
+            if matched_command:
+                if len(text) > 0:
+                    text += ' '
+                text += word
+            else:
+                if len(matched_command_str) > 0:
+                    matched_command_str += ' '
+                matched_command_str += word
+                for cmd in self.commands:
+                    if cmd.matches(matched_command_str):
+                        matched_command = cmd
+
+        if matched_command:
+            matched_command.do(text)
```

Comment:
This is a super fancy algorithm that chops what the user entered into a series of words, and tries to find a match on any of the existing commands or their aliases. This is how you can magically add an alias like "where can i go" and it will eventually match as it builds up the "command part" and the "rest of sentence" part of what the user typed in.

Reply:
Oh, and see `matched_command.do(text)` ?  THAT's how the code jumps from this part into the `do` function of whatever command is matched... it will feel a little magical now when you're creating new commands, you'll be like "wait but how do the code get here".  This is the exact spot where that link happens.